File: .config/wezterm/src/Fs.ts
import { Result, ok, err } from "@codethread/tstl-result";

const fileCache = new LuaMap<string, string>();

export const Fs = {
  readFile(
    file: string,
    opts: { noCache?: boolean } = {},
  ): Result<string, string> {
    const inCache = fileCache.get(file);
    if (!opts.noCache && inCache) {
      return ok(inCache);
    }
    const [f] = io.open(file, "rb");
    if (!f) return err(file);
    const content = f.read("*all" as any);
    if (typeof content != "string") throw new Error("ah");
    f.close();
    fileCache.set(file, content);
    return ok(content);
  },
};

File: .config/wezterm/_/fs.lua
local str = require '_.str'
local Fs = {}

local cache = {}

---Read a file from disk
---@param file string
---@param opts? { no_cache?: boolean }
---@return Str | nil
function Fs.readfile(file, opts)
	opts = opts or { no_cache = false }
	local cached = cache[file]
	if not opts.no_cache and cached then return cached end

	local f = io.open(file, 'rb')
	if not f then return nil end
	local content = f:read '*all'
	if type(content) ~= 'string' then return nil end
	f:close()
	local out = str(content)
	cache[file] = out
	return out
end

return Fs
